lib/bolt/transport/ssh/connection.rb

```diff
@@ -130,7 +130,11 @@ def connect
 
         def disconnect
           if @session && !@session.closed?
-            @session.close
+            begin
+              Timeout.timeout(@target['connect-timeout']) { @session.close }
```

Comment:
I am having a hard time reproducing the case described in https://github.com/puppetlabs/bolt/issues/666. Further conversation in slack revealed that there are other configuration details besides the simple `ProxyCommand`. Instead there is a combination of `ForceCommand` and `Command*`. I think that between the unsupported net-ssh `Control*` settings and the proprietary/custom jump host used by the reporter we get in to a situation where net-ssh hangs forever trying to close a socket. Here is the config I am using to try to repro this:
## local ~/.ssh/config
```
Host p71tjeuwlftc81l.delivery.puppetlabs.net
  IdentityFile ~/.ssh/id_rsa-acceptance
  User root
  StrictHostKeyChecking no
  HostName p71tjeuwlftc81l
  Port 22
  ProxyCommand ssh -q root@q0an0ruursqn43r.delivery.puppetlabs.net
  ControlMaster auto
  ControlPath ~/.ssh/sockets/%r@%h-%p
  ControlPersist 600
```
# Proxy ~/.ssh/config
```
Host p71tjeuwlftc81l.delivery.puppetlabs.net
  IdentityFile ~/.ssh/id_rsa-acceptance
  User root
  StrictHostKeyChecking no
  HostName p71tjeuwlftc81l
  Port 22
  ControlMaster auto
  ControlPath ~/.ssh/sockets/%r@%h-%p
  ControlPersist 600
```
# Proxy sshd config
```
root@q0an0ruursqn43r:~# cat /etc/ssh/sshd_config 
# File is managed by Puppet
ForceCommand /bin/test.sh  
AcceptEnv LANG LC_*
ChallengeResponseAuthentication no
GSSAPIAuthentication no
PermitRootLogin yes
PrintMotd no
Subsystem sftp /usr/lib/openssh/sftp-server
UseDNS no
UsePAM yes
X11Forwarding yes
```
## test.sh
```
#!/bin/bash
nc p71tjeuwlftc81l 22
```
Using the `ForceCommand` to execute a shell script mirrors what the reporter's system does. However even with this configuration I am unable to reproduce the hanging behavior. 

I put up this PR based on some comments in the source code of net-ssh and arbitrarily used the same `connect-timeout` for `disconnect-timeout`. If we can reproduce this or verify this patch works then perhaps we can introduce another setting and write some unit tests for this.